File: zig/ui/opengl/camera/shader.zig
const std = @import("std");
const gl = @import("glad.zig");

const glm = @cImport({
    @cInclude("cglm/cglm.h");
});


const ShaderType = enum {
    VERTEX_SHADER, FRAGMENT_SHADER, PROGRAM
};

const ShaderError = error {
    COMPILE_SHADER_ERROR,
    LINK_PROGRAM_ERROR,
};


pub fn Shader(allocator: std.mem.Allocator) type {
    return struct {
        const Self = @This();
        id: u32,
        allocator: std.mem.Allocator = allocator,

        pub fn createProgram(vertex_path: []const u8, fragment_path: []const u8, io: std.Io) !Self {
            var buf: [1024]u8 = .{0} ** 1024;
            const v_f = try std.Io.Dir.cwd().openFile(io, vertex_path, .{});
            defer v_f.close(io);
            var vf_reader = v_f.readerStreaming(io, &buf);
            const f_size = try v_f.length(io);
            const vertex_source = try vf_reader.interface.readAlloc(allocator, f_size);
            std.debug.print("{s}\n", .{vertex_source});
            defer allocator.free(vertex_source);

            const v_s = gl.glCreateShader(gl.GL_VERTEX_SHADER);
            gl.glShaderSource(v_s, 1, @ptrCast(&vertex_source), null);
            gl.glCompileShader(v_s);
            if (!checkStatus(v_s, .VERTEX_SHADER)) {
                return ShaderError.COMPILE_SHADER_ERROR;
            }

            const f_f = try std.Io.Dir.cwd().openFile(io, fragment_path, .{});
            defer f_f.close(io);
            var ff_reader = f_f.readerStreaming(io, &buf);
            const ff_size = try f_f.length(io);
            const fragment_source = try ff_reader.interface.readAlloc(allocator, ff_size);
            std.debug.print("{s}\n", .{fragment_source});
            defer allocator.free(fragment_source);

            const f_s = gl.glCreateShader(gl.GL_FRAGMENT_SHADER);
            gl.glShaderSource(f_s, 1, @ptrCast(&fragment_source), null);
            gl.glCompileShader(f_s);
            if (!checkStatus(f_s, .FRAGMENT_SHADER)) {
                return ShaderError.COMPILE_SHADER_ERROR;
            }

            const prog = gl.glCreateProgram();
            gl.glAttachShader(prog, v_s);
            gl.glAttachShader(prog, f_s);
            gl.glLinkProgram(prog);
            if (!checkStatus(prog, .PROGRAM)) {
                return ShaderError.LINK_PROGRAM_ERROR;
            }
            gl.glDeleteShader(v_s);
            gl.glDeleteShader(f_s);

            return .{.id = prog, .allocator = allocator};
        }

        fn checkStatus(element: u32, shader_type: ShaderType) bool {
            switch(shader_type) {
                .VERTEX_SHADER => {
                    var status: i32 = 0;
                    gl.glGetShaderiv(element, gl.GL_COMPILE_STATUS, @ptrCast(&status));
                    if (status == @as(i32, @intCast(gl.GL_FALSE))) {
                        var buf: [512]u8 = .{0} ** 512;
                        gl.glGetShaderInfoLog(element, 512, null, @ptrCast(&buf));
                        std.debug.print("Compile VERTEX shader error: {s}\n", .{&buf});
                        return false;
                    }

                },
                    .FRAGMENT_SHADER => {
                        var status: i32 = 0;
                        gl.glGetShaderiv(element, gl.GL_COMPILE_STATUS, @ptrCast(&status));
                        if (status == @as(i32, @intCast(gl.GL_FALSE))) {
                            var buf: [512]u8 = .{0} ** 512;
                            gl.glGetShaderInfoLog(element, 512, null, @ptrCast(&buf));
                            std.debug.print("Compile FRAGMENT shader error: {s}\n", .{&buf});
                            return false;
                        }
                    },
                    .PROGRAM => {
                        var status: i32 = 0;
                        gl.glGetProgramiv(element, gl.GL_LINK_STATUS, @ptrCast(&status));
                        if (status == @as(i32, @intCast(gl.GL_FALSE))) {
                            var buf: [512]u8 = .{0} ** 512;
                            gl.glGetProgramInfoLog(element, 512, null, @ptrCast(&buf));
                            std.debug.print("Progarm link shader error: {s}\n", .{&buf});
                            return false;
                        }
                    },
                }
            return true;
        }

        fn uniformLocation(self: *Self, uniform_name: []const u8) i32 {
            const loc = gl.glGetUniformLocation(self.id, @ptrCast(uniform_name));
            if (loc == -1) {
                std.debug.print("Can't find location of uniform {s}\n", .{uniform_name});
                return -1;
            }
            return loc;
        }

        pub fn use(self: *Self) void {
            gl.glad_glUseProgram.?(self.id);
        }
        pub fn destroy(self: *Self) void {
            gl.glad_glDeleteProgram.?(self.id);
        }

        pub fn setFloat(self: *Self, uniform_name: []const u8, val: f32) void {
            const loc = self.uniformLocation(uniform_name);
            if (loc == -1) return;
            gl.glUniform1f(loc, val);
        }

        pub fn setInt(self: *Self, uniform_name: []const u8, val: i32) void {
            const loc = self.uniformLocation(uniform_name);
            if (loc == -1) return;
            gl.glUniform1i(loc, val);
        }

        pub fn setVec2(self: *Self, uniform_name: []const u8, val: glm.vec2) void {
            const loc = self.uniformLocation(uniform_name);
            if (loc == -1) return;
            gl.glUniform2fv(loc, 1, val);
        }
        pub fn setVec3(self: *Self, uniform_name: []const u8, val: glm.vec3) void {
            const loc = self.uniformLocation(uniform_name);
            if (loc == -1) return;
            gl.glUniform3fv(loc, 1, val);
        }
        pub fn setVec4(self: *Self, uniform_name: []const u8, val: glm.vec4) void {
            const loc = self.uniformLocation(uniform_name);
            if (loc == -1) return;
            gl.glUniform4fv(loc, 1, val);
        }
        pub fn setMat4(self: *Self, uniform_name: []const u8, val: glm.mat4) void {
            const loc = self.uniformLocation(uniform_name);
            if (loc == -1) return;
            gl.glUniformMatrix4fv(loc, 1, gl.GL_FALSE, val);
        }

    };
}

File: c/ui/learnopengl/getting_started/test/shader.c
#include "shader.h"
#include <cstdext/io/logger.h>
#include <cstdext/io/reader.h>
#include <glad/glad.h>


static bool programCheckStatus(u32 element, u32 type) {

  switch(type) {
  case GL_VERTEX_SHADER: {
    i32 status;
    glGetShaderiv(element, GL_COMPILE_STATUS, &status);
    if (status == GL_FALSE) {
      i8 buf[512] = {0};
      glGetShaderInfoLog(element, 512, null, buf);
      log(ERROR, "Compile VERTEX shader error: %s",buf);
      glDeleteShader(element);
      return false;
    }
  } break;
  case GL_FRAGMENT_SHADER: {
    i32 status;
    glGetShaderiv(element, GL_COMPILE_STATUS, &status);
    if (status == GL_FALSE) {
      i8 buf[512] = {0};
      glGetShaderInfoLog(element, 512, null, buf);
      log(ERROR, "Compile FRAGMENT shader error: %s", buf);
      glDeleteShader(element);
      return false;
    }
  } break;
  case 0: {
    i32 status;
    glGetProgramiv(element, GL_LINK_STATUS, &status);
    if (status == GL_FALSE) {
      i8 buf[512] = {0};
      glGetProgramInfoLog(element, 512, null, buf);
      log(ERROR, "Link program error: %s", buf);
      return false;
    }
  } break;
  }

  return true;
}

static u32 programCompileShader(str shader_path, u32 shader_type) {
 str shader_source = readEntyreFile(shader_path);
  if (shader_source == null) {
    log(ERROR, "readEntyreFile error");
    return 0;
  }
  
  u32 shader = glCreateShader(shader_type);
  glShaderSource(shader, 1, (const i8 **)&shader_source, null);
  glCompileShader(shader);
  if (!programCheckStatus(shader, shader_type)) {
    glDeleteShader(shader);
    DEALLOC(shader_source);
    return 0;
  }

  DEALLOC(shader_source);
  return shader;
}


Program programCreate(str vertex_path, str fragment_path) {
  u32 v_s = programCompileShader(vertex_path, GL_VERTEX_SHADER);
  if (v_s == 0) {
    return 0;
  }
  u32 f_s = programCompileShader(fragment_path, GL_FRAGMENT_SHADER);
  if (f_s == 0) {
    return 0;
  }

  u32 prog = glCreateProgram();
  glAttachShader(prog, v_s);
  glAttachShader(prog, f_s);
  glLinkProgram(prog);
  glDeleteShader(v_s);
  glDeleteShader(f_s);
  if (!programCheckStatus(prog, 0)) {
    glDeleteProgram(prog);
    return 0;
  }

  return prog;
}
void programUse(Program p) {
  glUseProgram(p);
}

static i32 programGetUniformLocation(Program p, str uniform_name) {
  i32 loc = glGetUniformLocation(p, uniform_name);
  if (loc == -1) {
    log(ERROR, "Can't find location of uniform %s", uniform_name);
    return -1;
  }
  return loc;
}


void programSetUniformInt(Program p, str uniform_name, i32 val){
  i32 loc = programGetUniformLocation(p, uniform_name);
  glUniform1i(loc, val);
}
void programSetUniformFloat(Program p, str uniform_name, f32 val){
  i32 loc = programGetUniformLocation(p, uniform_name);
  glUniform1f(loc, val);
}
void programSetUniformVec4(Program p, str uniform_name, vec4 val){
  i32 loc = programGetUniformLocation(p, uniform_name);
  glUniform4fv(loc, 1, val);
}
void programSetUniformMat4(Program p, str uniform_name, mat4 val) {
  i32 loc = programGetUniformLocation(p, uniform_name);
}
